### DISCUSSION TITLE
ShouldApply

### DISCUSSION THREAD

**Discussion by danschappel** [idaholab/moose#28962] [Q&A Meshing]
Does the ShouldApply() clause still work with DirichletBCs? It still works with PenaltyDirichletBCs. However, when using the new computeQpValue() method DirichletBCs seem to have an issue of trying to both set and not set the boundary condition when should apply is false, and it produces a displacement that is neither 0 (the value if the BC is disabled) or the intended value when the BC is active. If you use the older computeQpResidual() it has convergence issues that seem similar to the results from setting use_preset = false.

**Comment by GiudGiud** [comment]
Hello

It should. `shouldApply` is defined in the BoundaryCondition parent class. 

>  and it produces a displacement that is neither 0 (the value if the BC is disabled) 

Why would it be 0 if the BC is disabled? 0 could be imposed with a DirichletBC but how do you know it is 0 if it's not imposed?

>  If you use the older computeQpResidual() it has convergence issues that seem similar to the results from setting use_preset = false.

ComputeQpResidual is set by the DirichletBCBase to (u - value). What are you modifying it to?

**Reply by danschappel** [reply]
That BC is the only source of displacement in the simulation at present. When the BC is removed from the input file the displacement is zero and there are no other BCs on the surface. 

I've left it at _u[_qp] - value at the moment, but it still has this issue. However, it is good to know shouldapply is supposed to still work.

**Comment by danschappel** [comment]
Ok I think I got it. When using a DirichletBC with shouldapply and computeQpValue(), at the beginning of each timestep it is trying to set the surface to the value if the BC is active and then work towards to an inactive condition. If the value is similar to what would happen if there was no BC this works, but if not then it's a problem.
